File: middle-of-the-linked-list/Solution876.java
// https://leetcode.com/problems/middle-of-the-linked-list/

class Solution876 {

    // T: O(n), S: O(1)
    public ListNode middleNode(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    public static void main(String[] args) {
        var sol = new Solution876();
        // 1,2,3,4,5
        var head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(3);
        head1.next.next.next = new ListNode(4);
        head1.next.next.next.next = new ListNode(5);
        System.out.println(head1);
        System.out.println(sol.middleNode(head1).val); // 3
    }

}

File: middle-of-the-linked-list/solution876.py
# https://leetcode.com/problems/middle-of-the-linked-list/

from typing import Optional


class ListNode:

    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next

    def __repr__(self):
        return f"{self.val} -> {self.next.__repr__() if self.next else 'None'}"


class Solution876:

    # O(n), S: O(1)
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow


def main() -> None:
    sol = Solution876()
    head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
    print(sol.middleNode(head1))


if __name__ == "__main__":
    main()
